p4include/psa.p4

```diff
@@ -0,0 +1,672 @@
+  ///
+  /// @param data A header definition that can be added to the set of
+  ///             packet headers.
+  /// @param port The input port at which the packet will be resubmitted.
+  void resubmit<T>(in T data, in PortId_t port);
+
+  /// Recirculate operation: send a post deparse packet to the ingress
+  ///                        port with additional data appended
+  ///
+  /// This operation is intended for recursive packet processing.
+  /// \TODO: needs both source and destination data
+  ///
+  /// @param data A header definition that can be added to the set of
+  ///             packet headers.
+  /// @param port The input port at which the packet will be resubmitted.
+  void recirculate<T>(in T data, in PortId_t port);
+
+  /// Truncate operation
+  ///
+  /// Truncate the outgoing packet to the specified length
+  ///
+  /// @param length packet length
+  void truncate(in bit<32> length);
+
+  /*
+  @ControlPlaneAPI
+  {
+  // TBD
+  }
+  */
+}
+
+/**
+ * The BufferingQueueingEngine extern represents the the other
+ * non-programmable part of the PSA pipeline (after Egress).
+ *
+ * Even though the BQE can not be programmed using P4, it can be
+ * configured both directly using control plane APIs and by setting
+ * intrinsic metadata. In this specification we opt to define the
+ * operations available in the BQE as method invocations. A target
+ * backend is responsible for mapping the BQE extern APIs to the
+ * appropriate mechanisms for performing these operations in the
+ * hardware.
+ *
+ * The BQE is instantiated by the architecture and a P4 program can
+ * use it directly. It is an error to instantiate the BQE multiple
+ * times.  The BQE is made available to the Egress programmable block
+ * using the same mechanism as packet_in. A corresponding Packet
+ * Replication Engine (PRE) extern is defined for the Igress pipeline
+ * (@see PacketReplicationEngine).
+ *
+ * Note: some of these operations may not be implemented as primitive
+ * operations on a certain target. However, All of the operations can
+ * be implemented as a combination of other operations. Applications
+ * that rely on non-primitive operations may incur significant
+ * performance penalties, however, they should be functionally
+ * correct.
+ *
+ * The ordering semantics of multiple calls to BQE APIs is identical
+ * to the semantics ordering of PRE invocations, for the subset of
+ * functions supported in the BQE.
+ */
+extern BufferingQueueingEngine {
+
+  // BufferingQueueingEngine(); /// No constructor. BQE is instantiated
+                                /// by the architecture.
+
+  /// Unicast operation: sends packet to port
+  ///
+  /// Targets may implement this operation by setting the appropriate
+  /// intrinsic metadata or through some other mechanism of
+  /// configuring the BQE.
+  ///
+  /// @param port The output port. If the port is PORT_CPU the packet
+  ///             will be sent to CPU
+  void send_to_port (in PortId_t port);
```

Comment:
Sending to a port AFTER egress is typically not possible or extremely expensive on most modern architectures.